### PR TITLE
Add train script that runs on the TPU VM

### PR DESCRIPTION
**Pull request by nostalgebraist**
Adds a script for training on the TPU VM, similar to the existing `device_sample.py`, etc.

My use case: I'm fine-tuning the model on a single TPU v3-8.  I initially got this working with the existing `train.py`, but

- I had to work around its expectations about file structure e.g. `meta.json`
- this way has lots of complexity that is needless here: using `ray`, using an extra VM outside the TPU, the `TPUCluster` wrapper, etc.

Note: I slapped this together hastily and it's not perfect.  I haven't yet verified it works on this branch, just on a personal dev branch with other changes.

### Comments

**Step**

- Compared to `train.py`, I changed when `step` is incremented, to try to make the values in the checkpoint file names match the exact number of gradient steps that have been run.
  - I increment the step after compiling model fn
  - Inside the loop, we do (update weights) --> (increment step) --> (save checkpoint if necessary), whereas train.py saves in between the update and the save
- As a result of this, the val steps are now incremented by one, i.e. we'll do the first val at step 1 (because we took step 1 compiling model_fn)

**Data loader**

- The changes to the TFRecord reader class cover 2 issues:
  - On my small finetuning dataset, I'm doing multiple epochs.  After save/load, the iterator would recurse infinitely because every file was in the `used` list that was loaded.
    - Fixed with logic to discard `used` entries from earlier epochs
  - I added `reset()` method to try to make val data sampling deterministic.  (I have a tiny val dataset)

**Script**

- Does fine-tuning if `tune-model-path` is provided, otherwise trains from scratch (just for completeness)
- Compared to `train.py`, adds some more wandb metrics, print statements, etc.  You can take or leave these, they're just what I wanted for my project
- `save()` is mostly a copy-paste of `TPUCluster.save` which has some useful saving logic.  Ideally this logic would be defined only once, but I really wanted to avoid breaking `TPUCluster`

### CODE REVIEW

**Comment by nostalgebraist** [comment]
Noticed an OOM trying to run this, closed PR until I find out if it's a persistent issue

**Comment by nostalgebraist** [comment]
> Noticed an OOM trying to run this, closed PR until I find out if it's a persistent issue

The OOM is reproducible, it happens in the second train step, after doing a full eval pass.  No idea what's going on, the same config works with `train.py` and ray

**Comment by nostalgebraist** [comment]
> > Noticed an OOM trying to run this, closed PR until I find out if it's a persistent issue
> 
> The OOM is reproducible, it happens in the second train step, after doing a full eval pass. No idea what's going on, the same config works with `train.py` and ray

Resolved with help from Discord, re-opened as #16
